### ISSUE TITLE
Newline characters are introduced into long lines, causing bugs/errors.  Lines >4096 chars are truncated.

### ISSUE THREAD

**Issue by phil-s** [bobthecow/psysh#496]
I believe I'm seeing 1024-byte buffering at play here and, when an input line exceeds that length, newlines are introduced into the PHP input (at every 1024-character boundary), which can cause bugs and errors.

This might be connected to #155 ?

For example, with exactly 1024 dots, plus the 2 double-quote characters wrapping them:

    >>> "................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................"
    ... => """
       ..............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................\n
       ..
       """

(Scrolling to the end of the line shows the introduced newline, and we see how two of the dots have been pushed to the following line.)

We can easily see how this can cause bugs:

For instance, the following array key is corrupted:

    >>> $x["................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................"] = "foo";
    ... => "foo"

    >>> $x
    => [
         "...........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................\n....." => "foo",
       ]

And here the newline corrupts the code and causes a parse error:

    >>> $x = "....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................."; $x;
    ... PHP Parse error: Syntax error, unexpected T_STRING, expecting T_VARIABLE or '{' or '$' on line 2

Due to the newline being inserted between the `$` and the `x` of the latter instance of `$x`, which turns that code into this:

    >>> $
    x
    ... PHP Parse error: Syntax error, unexpected T_STRING, expecting T_VARIABLE or '{' or '$' on line 2

More generally, using PHP input where the newlines have been removed will mean that boundaries could occur at *any arbitrary position* in the code, so all manner of different failures may result.

Can we eliminate these additional newlines?

(I found this on account of my own workaround for #490 in which I am automatically eliminating newlines in order to prevent PsySH commands being executed, which is now triggering this new bug for larger inputs :/ )

**Comment by bobthecow** [comment]
I'm not entirely sure what's up here, but it seems to have something to do with limitations in readline :-/

**Comment by phil-s** [comment]
This happens with 'useReadline' => false in the config file.  Would that not exclude readline from the equation?

**Comment by phil-s** [comment]
Modifying the `1024` in `Psy\Readline\Transient::readline()` affects this for me.  If I change that to `10` then I can see, for example:

    >>> '1234567890'
    ... => """
       12345678\n
       90
       """

Regardless of the `Readline` implementation being used, `Psy\Shell::getInput()` gets the buffered input string with `$input = $this->readline();` and calls `$this->addCode($input);` for each string, and `addCode` adds the input string to an *array* and then 'cleans' the array:

    $this->codeBuffer[] = $silent ? new SilentInput($code) : $code;
    $this->code         = $this->cleaner->clean($this->codeBuffer, $this->config->requireSemicolons());

The newlines are introduced by  `Psy\CodeCleaner::clean()` with `implode(PHP_EOL` here:

    $stmts = $this->parse('<?php ' . \implode(PHP_EOL, $codeLines) . PHP_EOL, $requireSemicolons);

**Comment by phil-s** [comment]
More related problems:

`Transient::readline()` trims the right-hand side of the input, so necessary mid-statement whitespace can be removed.

`CodeCleaner::clean()` calls `parse()` on whatever it got from the buffer, which leads to parsing errors.

With a 10-character buffer size, and `var_dump($code);` at the top of `CodeCleaner::parse()` we get this kind of behaviour:

    >>> $foo = ['1','2','3'];
    string(16) "<?php $foo = ['
    "
    PHP Parse error: Syntax error, unexpected T_ENCAPSED_AND_WHITESPACE, expecting ']' on line 1
    >>> string(16) "<?php 1','2','3
    "
    PHP Parse error: Syntax error, unexpected T_CONSTANT_ENCAPSED_STRING on line 1
    >>> string(11) "<?php '];

    "

Which still thinks it's in the middle of a string, so we get out of that with:

    ... ';
    string(14) "<?php '];
    ';

    "
    >>>return '];
    ';<<<=> "];\n"
    >>>

**Comment by phil-s** [comment]
Here is an initial improvement:

In `Shell::getInput()` change this:

            $input = $this->readline();

to:

            $input = '';
            do {
                $line = $this->readline();
                $input .= $line;
            }
            while (substr($line, -1) !== "\n");

**Comment by bobthecow** [comment]
Thanks for digging into this further.

Something like that should work for the `Transient` readline implementation, but it won't work for `GNUReadline` or `Libedit`, because those don't end with `\n` when they're finished.

So I'd prefer to push it down into `Transient` rather than changing `Shell::getInput`:

```diff
diff --git a/src/Readline/Transient.php b/src/Readline/Transient.php
index e238fdf..cad1070 100644
--- a/src/Readline/Transient.php
+++ b/src/Readline/Transient.php
@@ -106,7 +106,13 @@ class Transient implements Readline
     {
         echo $prompt;
 
-        return \rtrim(\fgets($this->getStdin(), 1024));
+        $input = '';
+        do {
+            $line = \fgets($this->getStdin(), 1024);
+            $input .= $line;
+        } while (substr($line, -1) !== "\n");
+
+        return \rtrim($input);
     }
```

Note that I'm still doing the `\rtrim` to remove the final `\n`, to match what `\readline` returns.

**Comment by phil-s** [comment]
Looking at the behaviour of `fgets`, is there a reason not to do this: ?

    -        return \rtrim(\fgets($this->getStdin(), 1024));
    +        return \rtrim(\fgets($this->getStdin()));

I'm thinking there may be a reason, on the basis that a limit *was* imposed in the first place, but I don't know what the reason would be.

Without that length parameter, it should keep reading until it either has a line, or EOF.

>  Reading ends when length - 1 bytes have been read, or a newline (which is included in the return value), or an EOF (whichever comes first). If no length is specified, it will keep reading from the stream until it reaches the end of the line.
>
>  Note:
>
>  Until PHP 4.3.0, omitting it would assume 1024 as the line length. If the majority of the lines in the file are all larger than 8KB, it is more resource efficient for your script to specify the maximum line length.

That note would mean that if psysh needs to retain compatibility with PHP 4.3 then we would still need to use the loop; but I think we could also still drop the explicit parameter?

**Comment by phil-s** [comment]
I'll add that once psysh was doing the right thing, I then found myself hitting a 4096 character limit, which I think is some kind of I/O buffering imposed by the OS on account of using a pseudo-terminal to communicate with psysh.  In this instance the excess was simply discarded, and there didn't appear to be anything that psysh could do about it.

Happily, in a real terminal (or at least in gnome-terminal specifically) I'm not encountering that issue.

I found that using a pipe instead of a pty eliminated that 4K issue, and seems to be working well (partly thanks to being able to set `'colorMode' => \Psy\Configuration::COLOR_MODE_FORCED,` to still obtain coloured output!)

I just threw a "......." input string with 600,000 dots, and got all 600,000 back in the output :)

I've also tested installing the `php5-readline` Debian package and trying `Psy\Readline\GNUReadline` and this is where I first noticed that 4K buffer with the pty, but I'm guessing it was again on account of the pty.  Using a pipe didn't play nicely with `GNUReadline`, so I couldn't test that alternative; but when using psysh in gnome-terminal with `'useReadline' => true,` configured, I could successfully repeat that 600,000 char string test.  Although I now see that `\Psy\Configuration::getReadline()` tells me I'm actually using `Psy\Readline\Libedit` there.

In any case, my general impression is that for normal use-cases the fix to `Transient` is sufficient, and that configurations where it doesn't work are probably not psysh's problem!

**Comment by phil-s** [comment]
Oh, I've not tried `HoaConsole` (and the underlying library doesn't seem to have been installed by composer), and you didn't mention that one in your last comment, so I'm just mentioning it now in case that might need a similar adjustment to `Transient`.

**Comment by phil-s** [comment]
Ah, I spoke too soon.  I had failed to test `Transient` under `gnome-terminal`, and doing so now I see that I again encounter a 4K limit.

So it wasn't that a pseudo-terminal was behaving differently to a real terminal, but rather it seems to me that the `Transient` readline is triggering an I/O buffering issue with terminals in general?

**Comment by phil-s** [comment]
I've now tried `HoaConsole` which (a) seems quirky, but (b) seemingly does not have the same issue as `Transient`, so that seemed like it should provide insight into the issue.

I could see it was using `stty -icanon` so I tried that in my terminal.  That appeared to solve the 4K issue, but I quickly noticed that Ctrl-d also didn't quit the REPL, and indeed `-canon` disables all line editing features for the terminal.  I believe that the 4K buffer is therefore the line-editing buffer, and it discards any input larger than that maximum line size.

http://www.linusakesson.net/programming/tty/ says:

> Line editing. Most users make mistakes while typing, so a backspace key is often useful. This could of course be implemented by the applications themselves, but in accordance with the UNIX design philosophy, applications should be kept as simple as possible. So as a convenience, the operating system provides an editing buffer and some rudimentary editing commands (backspace, erase word, clear line, reprint), which are enabled by default inside the line discipline. Advanced applications may disable these features by putting the line discipline in raw mode instead of the default cooked (or canonical) mode. Most interactive applications (editors, mail user agents, shells, all programs relying on curses or readline) run in raw mode, and handle all the line editing commands themselves. The line discipline also contains options for character echoing and automatic conversion between carriage returns and linefeeds. Think of it as a primitive kernel-level sed(1), if you like.

**Comment by bobthecow** [comment]
>I'm thinking there may be a reason, on the basis that a limit was imposed in the first place, but I don't know what the reason would be.

I'm not sure either 🙂

The length parameter on `fgets` [dates back to the initial commit in this repo](https://github.com/bobthecow/psysh/blob/495081083cfcc9cda29d2787d13fcf7a683b92de/src/Psy/Shell.php#L334). If I had to guess, I imagine it's there because it _was_ to support PHP 4.3 in the script I used as a predecessor to this project _long_ before open sourcing it.

>I'll add that once psysh was doing the right thing, I then found myself hitting a 4096 character limit… In this instance the excess was simply discarded, and there didn't appear to be anything that psysh could do about it.

This was the behavior I was noticing in my investigations back in May, when I said:

>I'm not entirely sure what's up here, but it seems to have something to do with limitations in readline :-/

It makes a lot of sense that this is a limitation in tty / line editing, and seems consistent with what I observed. 

>I've now tried `HoaConsole` which (a) seems quirky, but (b) seemingly does not have the same issue as Transient, so that seemed like it should provide insight into the issue.
>
>I could see it was using `stty -icanon` so I tried that in my terminal.

Yep, that's exactly where I went when investigating this in May :) 

`HoaConsole` is a little weird. It basically reimplements readline, in userland PHP, on top of lower level raw mode tty input. At one point I considered going a similar route with the `Transient` implementation, but we ended up adding support for `HoaConsole` based readline instead. I figure there's no need for two projects to take on that mess 😛 

>I've updated the description of `useReadline`…

Perfect, thank you. I've removed the limit on `fgets` in dc01559782f3f97b4b1bfffdcfff505992c5083c as well.

**Comment by bobthecow** [comment]
Thank you so much for being so thorough with this. With the documentation change and raising the limit on `fgets`, there's probably not much more we can do here, right?

**Comment by phil-s** [comment]
Agreed, I think we're done here.

I almost hate it when that amount of time and effort boils down to a one-line code change, but it's a good thing really :)

I concur with keeping the `Transient` behaviour distinct from what `HoaConsole` is doing.  When I started to look at that I was thinking perhaps there was some code or an approach that Transient could appropriate; but once I realised what HoaConsole was doing I decided that things were good just as they were.
